### ISSUE TITLE
Still getting hull errors

### ISSUE THREAD

**Issue by mtsamis** [dyn4j/dyn4j#83]
Unfortunately I'm still getting various errors after 1134b731412055bfa1f6e7e7e49d4f6d6874baf4. From what I saw, most of them have to do with colinear points on the hull. At this point I dunno if we can remove some of the algorithms altogether if they prove to be much trouble to implement correctly, or re-implement them from scratch.

**`Graham scan`**
```java
Counter example for GrahamScan
for point cloud: new Vector2[] {
5.916275853509346, -4.228267720344762
8.31483976082672, -0.3807196367883092
3.9941738969349405, -0.491971233546733
-5.952110964171484, -0.7480752942332325
};

Result is: new Vector2[] {
5.916275853509346, -4.228267720344762
8.31483976082672, -0.3807196367883092
3.9941738969349405, -0.491971233546733
};

At least one missing point:
(-5.952110964171484, -0.7480752942332325)
```

**`Gift wrap`**
```java
Counter example for GiftWrap
for point cloud: new Vector2[] {
-0.9025337983824699, 4.56709308364953
-5.5168621708920345, 0.34366552069341916
-2.400927400987851, 3.19563523962121
-9.419896312210547, 3.19563523962121
};

Result is: new Vector2[] {
-9.419896312210547, 3.19563523962121
-5.5168621708920345, 0.34366552069341916
-2.400927400987851, 3.19563523962121
};

At least one missing point:
(-0.9025337983824699, 4.56709308364953)
```

Also this one causes infinite loop:
```java
for point cloud: new Vector2[] {
-0.33826889474805055, 8.329321811558497
-3.5586156659982215, -3.467244912905423
-3.5586156659982215, -4.566140779700733
-3.5586156659982215, -3.05702346750299
1.1178446483487536, -3.05702346750299
```

**`Divide and Conquer`**
```java
Error for DivideAndConquer
for point cloud: new Vector2[] {
-5.214810023866061, -5.581528163221621
-3.2956195481849493, 6.700146933201903
2.159226322162535, -2.2353877725618476
4.84788802330902, -6.921113359457114
4.84788802330902, -6.921113359457114
};

Unexpected exception occured: 
java.lang.NullPointerException
	at org.dyn4j.geometry.hull.LinkedVertexHull.getEdge(LinkedVertexHull.java:366)
	at org.dyn4j.geometry.hull.LinkedVertexHull.mergeHulls(LinkedVertexHull.java:236)
	at org.dyn4j.geometry.hull.LinkedVertexHull.merge(LinkedVertexHull.java:109)
	at org.dyn4j.geometry.hull.DivideAndConquer.divide(DivideAndConquer.java:95)
	at org.dyn4j.geometry.hull.DivideAndConquer.generate(DivideAndConquer.java:66)
	at mainPackage.AppMain.v(AppMain.java:94)
	at mainPackage.AppMain.<init>(AppMain.java:231)
	at mainPackage.AppMain.main(AppMain.java:610)
```

**Comment by wnbittle** [comment]
I'm not ready to give up on them just yet.  I figured we'd find more.  Let me spend some time on these as well.

**Comment by wnbittle** [comment]
I won't close this one so quickly :).  I just committed another set of fixes for these failure cases.  Feel free to continue testing to see if you can break them again.

**Comment by mtsamis** [comment]
Yeah, don't close this yet, I have fresh counterexamples for you ;)

**`DivideAndConquer`**
1) After the last commit DivideAndConquer lasts longer but has problems with some shapes of this sort. From what I saw all of them are of the form: A lot of colinear points in one side and one (or more like in the example) other ones.
```java
Counter example for DivideAndConquer
for point cloud: new Vector2[] {
new Vector2(-14.824543255380023, 15.85178543840377), 
new Vector2(-7.137129280540021, 9.172487271304316), 
new Vector2(-2.4267381389973397, 5.0798096789437), 
new Vector2(3.8090704585325725, -0.3382444381082488), 
new Vector2(5.468709158244611, 3.749237201781721), 
new Vector2(8.603427044732271, -4.503876374079454), 
new Vector2(18.28323008135373, -1.7637916875406638), 
};

Result is: new Vector2[] {
new Vector2(-7.137129280540021, 9.172487271304316), 
new Vector2(8.603427044732271, -4.503876374079454), 
new Vector2(18.28323008135373, -1.7637916875406638), 
};

At least one missing point:
(-14.824543255380023, 15.85178543840377)
```

2) For some more rare scenarios I can get it to fall into an infinite loop as well. Here's one
```java
for point cloud: new Vector2[] {
new Vector2(6.139549211208822, -9.442599680596068), 
new Vector2(-9.48574757482028, -4.145392302141662), 
new Vector2(-6.546174205893472, -4.145392302141662), 
new Vector2(-6.546174205893472, -4.145392302141662), 
new Vector2(-6.546174205893472, 2.5290473414225345), 
new Vector2(5.853898089437757, -1.0973436367120737), 
new Vector2(16.934101639148516, -4.3377400821178655), 
new Vector2(-4.074289990801132, -9.18993634425481), 
new Vector2(-5.939730395013038, 6.555842975889754), 
new Vector2(-3.699493009315612, -12.353516669309741), 
new Vector2(-6.400662737169788, 10.44647326605104), 
new Vector2(-2.3765210797735703, -23.520436825151126), 
new Vector2(-5.499521085160579, 2.8401313230491), 
new Vector2(2.6238153563780715, -65.7271960828315), 
new Vector2(-5.737338842922952, -5.240574938670375), 
};
```

**`GiftWrap`**
1) Still has issues with infinite loops. The first easier case is when there are x points that are (almost?) colinear. I understand this is somewhat malformed, but in all cases it shouldn't go in endless loop. Here's small reproduction case:
```java
for point cloud: new Vector2[] {
new Vector2(6.85418688038985, 3.60407755644377), 
new Vector2(1.3880441403354804, -7.731358359315308), 
new Vector2(-2.4841551580755885, -15.761347528560703), 
};
```
2) But it also goes into endless loop for other cases (they seem to still have those colinear points in them). So here's a less trivial example:
```java
for point cloud: new Vector2[] {
new Vector2(0.5218441132359297, -5.277149829971936), 
new Vector2(-9.996100987411396, 4.890847254302056), 
new Vector2(-3.1570218115673954, 4.890847254302056), 
new Vector2(-10.429214649857256, 5.309550567046834), 
new Vector2(3.048698548571761, 4.890847254302056), 
new Vector2(-1.4338079622141695, 4.79163160321358), 
new Vector2(-6.247975181961185, 7.425546117235971), 
new Vector2(3.559910732505469, 5.227841167113247), 
new Vector2(2.154116363687857, -7.027521301833401), 
new Vector2(0.1629631444656927, -24.38589531018993), 
};
```

**`GrahamScan`**
1) There seem to be some simple-ish cases that produce wrong output, like the following:
```java
Counter example for GrahamScan
for point cloud: new Vector2[] {
-11.546716981773478, 0.7207966448933689
1.7439140529994224, 8.460062386950122
-5.849229446365878, 4.038499918967357
-7.520640194675368, 4.43799040250121
};

Result is: new Vector2[] {
-11.546716981773478, 0.7207966448933689
-5.849229446365878, 4.038499918967357
-7.520640194675368, 4.43799040250121
};

At least one missing point:
(1.7439140529994224, 8.460062386950122)
```
2) Although this probably is the same class of problems with the above here's a more complex example, with multiple points missing from the hull:
```java
Counter example for GrahamScan
for point cloud: new Vector2[] {
new Vector2(18.02823901125023, -48.563856285711815),
new Vector2(5.837366515869737, 1.9642225141622771),
new Vector2(-1.6624555303018322, 33.04908534720289),
new Vector2(11.283104362182621, -20.60698156710606),
new Vector2(8.335476689183658, -9.303559083729326),
new Vector2(2.8185940934160603, 4.173994413634454),
new Vector2(2.8185940934160603, 3.9614676010893923),
new Vector2(2.8185940934160603, 2.4457852260040056),
new Vector2(-0.467595291720615, 8.01876668182382),
new Vector2(-4.120906637353855, 17.224889549524125),
new Vector2(2.4975302436899827, -2.85851098830896),
new Vector2(2.8185940934160603, -4.050896610684694),
new Vector2(-0.08457075085394539, -0.528762437876231),
new Vector2(-1.0692315002824904, 1.2037705703528156),
new Vector2(-0.6830502469109989, 0.011226796219898638),
new Vector2(-0.6830502469109989, 0.011226796219898638),
new Vector2(-0.6830502469109989, 0.011226796219898638),
new Vector2(-0.6830502469109989, 0.011226796219898638),
new Vector2(0.4189445731108843, -2.8498986645966395),
new Vector2(-4.124373444211207, 6.965938040682511),
new Vector2(-4.124373444211207, 6.965938040682511),
new Vector2(-4.554478357993627, 5.288053601795122),
new Vector2(-7.453410907207305, 9.950651102108008),
new Vector2(-5.065886265583872, 3.167462954859024),
new Vector2(-6.446692710201411, 2.254612493908348),
new Vector2(-6.996136032111744, -4.8364602156627035),
new Vector2(-4.120906637353855, -9.954262435143919),
};

Result is: new Vector2[] {
new Vector2(18.02823901125023, -48.563856285711815),
new Vector2(5.837366515869737, 1.9642225141622771),
new Vector2(-4.120906637353855, 17.224889549524125),
new Vector2(-7.453410907207305, 9.950651102108008),
new Vector2(-6.996136032111744, -4.8364602156627035),
new Vector2(-4.120906637353855, -9.954262435143919),
};

At least one missing point:
(-1.6624555303018322, 33.04908534720289)
```

**Comment by wnbittle** [comment]
Thanks for the additional test cases.  Let’s see if I can do anything with these.  Apparently colinearity is a major problem for the side of line tests these algorithms use.  Really hating finite precision right now.

**Comment by mtsamis** [comment]
Yeah floating points + geometry algorithms = trouble.

On a side note, I was thinking that for this same reason there should probably be a error/tolerance parameter to decomposition algorithms that is used when checking for such conditions as colinear points.

It is annoying to get points in the hull because they're barely colinear, but due to errors do not satisfy the test for colinearity. I believe this option is important because if I want to do it after the initial hull result it is computationally quite expensive and error prone.

**Comment by wnbittle** [comment]
Upon further investigation, I don't know if we'll be able to solve all the scenarios.  There's already some papers about how to make the side-of-line test robust and some github projects as well.  For example:
https://github.com/mourner/robust-predicates

I think 1,000 lines of code may be overkill.  I'm leaning towards just using an epsilon to determine near-colinearity and remove the vertices.  The only challenge will be with the automated tests - it won't know that we removed those for that reason so it will fail.

So far MonotoneChain has worked for all the examples I've tried.  Have you run into any issues with this algorithm?

**Comment by mtsamis** [comment]
Your linked repository (actually the work on which is based) is quite interesting.

My automated test btw (which I don't know if we really should include it in the tests or just keep it standalone) does not really has a problem with using some epsilon because it already uses some tolerance parameter to remove almost colinear points (because those did not give interesting bugs for result).

In my opinion, for the hull algorithms to be useful, they should be robust in a sense. Not about whether they will include some almost colinear point, but they should always 1) terminate 2) include points that are obviously in the hull. As in the above counterexamples I have made sure that the output is visibly wrong.

I would like to take a shot at making an implementation of the [robust side of line test](https://www.cs.cmu.edu/~quake/robust.html); Maybe with that we will be able to solve most if not all of those bugs. So if you can wait a bit and I'll get back with an implementation of that, I think it's worth it (given that I succeed in writing it properly).

MonotoneChain passed every test (and I ran some billions of them to be sure). It also is very good at doing point de-duplication and removing *unneeded* colinear points, with rare exceptions where it will include some of those (which is not wrong, and probably has to do with the side of line test being imprecise). But the good thing is it never gave output that is not a correct hull.

I'll get back to you when I have made progress with the implementation.

**Comment by wnbittle** [comment]
Go for it.  My only request is that it’s well documented and understandable.  The implementation in the linked project would not be acceptable as is.  It wasn’t clear what all the methods did, which arguments were mutable outside the function scope, etc.  I need to be able to understand code so I can support it :)

I’ll read through the source paper and see what I can understand.

**Comment by wnbittle** [comment]
Something else I was thinking about was that it would be fine to use a robust method for the hull generators, but I’d be cautious about using it in the engine proper.  We can evaluate as we learn more.

**Comment by mtsamis** [comment]
Yes, agreed. I was looking at the original implementation of the author in C. It is more complex but has quite more comments around what exactly is happening. I’ll see much I can understand it myself to begin with :)

And yeah, initially this is just for the hull algorithms; not to replace the implementation in the Segment class. We should evaluate the performance trade off after all. 

From what I saw in the code the implementation should be almost as fast for the ‘easy’ cases but will get slower as it has to deal with edge cases or in near colinearity where a lot of extra precision is needed. 
For me this essentially means that it’s ok to use where a lot of calls will be made e.g. the hull algorithms, because the cost will amortize but not for single calls where there can be an unexpected very slow call. 

This will probably take some time, but I’ll give it a good try.

**Comment by mtsamis** [comment]
@wnbittle here are some updates on the topic. I have made a draft implementation of the said algorithm for side of line testing that is quite readable. I need some time to fix some bugs it has, write tests and documentation. I'm trying to make the algorithm very understandable by trading some speed (maybe up to 2x for the hardest cases, but because the whole thing is complex I go for readability etc).

While I'm writing these, the question occurred to me "but will this new exact algorithm solve the hull algorithms bugs?". So I made a simple slow exact implementation with Java's BigDecimal and made the HullAlgorithms use it. It turns there are still other bugs that will need to be fixed.

- Good news: GiftWrap passes all the automated tests and seems to neither hang nor produce wrong output. Just by replacing the getLocation method.

- Not super good news: Both GrahamScan and DivideAndConquer still produce bad output. But hopefully, since the exact implementation didn't help, this probably means there are some 'straightforward' algorithmic bugs to fix and then the algorithms will function properly.

Do you want to check for those bugs while I'm writing the other implementation? I can provide counterexamples for the versions using the exact arithmetic for getLocation which is below. 
```java
public static double getLocation(Vector2 pa, Vector2 pb, Vector2 pc) {
    BigDecimal pax = new BigDecimal(pa.x);
    BigDecimal pay = new BigDecimal(pa.y);
    BigDecimal pbx = new BigDecimal(pb.x);
    BigDecimal pby = new BigDecimal(pb.y);
    BigDecimal pcx = new BigDecimal(pc.x);
    BigDecimal pcy = new BigDecimal(pc.y);
    
    BigDecimal result = pcx.subtract(pbx).multiply(pay.subtract(pby)).subtract(pax.subtract(pbx).multiply(pcy.subtract(pby)));
    double locExact = result.doubleValue();
    
    return locExact;
}
```

**Comment by mtsamis** [comment]
Hey William, I have (in a sense) finished my implementation of the robust side of line test etc.
Can you check the code in the [implementation in my branch](https://github.com/mtsamis/dyn4j/tree/robust_geometry) before I make a pull request?

I think I managed to make the code very readable and understandable (especially compared to that old unreadable C in the original work lol), up to dyn4j's standards, but I would like to have your opinion as well. This includes if the code/logic is clear enough but also if you like variable/class names and other 'cosmetic' stuff. I know there is a lot to read so no rush with it. When you are good with it, and I also make some minor final changes I'll go ahead with a PR.

The implementation is split in two parts: `CompoundDecimal` and `RobustGeometry`.

`CompoundDecimal` is a multi/variable precision decimal representation. You can think of it of being close to java's `BigDecimal` but 1) it's not full-blown arbitrary precision but rather has some lower/upper limits coming from the value range of `double` (this is not a problem for the said algorithm) 2) It does not have every operation implemented but rather what was needed to implement theside of line test. This can be filled with more in the future if needed but this will do for now (This is a reason I didn't make the class public) 3) It's much faster to operate on what we need

`RobustGeometry` contains what is necessary for the implementation of the adaptive side of line test using `double`s and `CompundDecimals`.

I also added a lot of useful (both standard and randomized) unit tests which showcase some properties and use cases of those classes.

There's also extensive javadoc and some comments which more or less explain all the bits, I think to a point where it's not even that important to read the original paper. Feel free to ask any questions here.
___

Sample speed results on uniform random vectors (should be a decent approximation of the expected slowdown in the algorithms that will be made to use this):
- Segment.getLocation: 1 time unit (baseline)
- RobustGeometry.getLocation: 1.25 - 1.75  time units (decent, adaptive speed)
- BigDecimal implementation: 1500.0 - 2500.0 time units (unusable, full precision)

Sample speed results on (almost) colinear random vectors i.e. the hardest case to solve (should be a decent approximation of the expected worst case slowdown):
- Segment.getLocation: 1 time unit
- RobustGeometry.getLocation: 10.0 - 15.0 time units
- BigDecimal implementation: 1500.0 - 2500.0 time units

Although this is not a full blown benchmark those are the approximate figures which are pleasing for me after all. This means that the new algorithm can safely be used and expect very minor slowdown. The worst case performance is good enough since the total runtime of Segment.getLocation is close to zero (4 products 3 sums).

**Comment by mtsamis** [comment]
I fixed all the (problematic) hull algorithms. They now pass all the randomized tests and what's more handle all the degenerate cases in a similar manner. In the process I also simplified and improved performance of some algorithms, especially that of GrahamScan & DivideAndConquer.

@wnbittle once we're done with the robust geometry stuff I explained above I can provide a PR with the new changes so we can finally close this topic (hopefully for good).

**Comment by wnbittle** [comment]
Finally getting to this.  I haven't run/tested the code, but this looks really good.  I do have a couple of comments/questions:

- **strictfp** - I read in the paper/site that x86 larger registers would round differently and could cause the algorithm to fail.  I saw that you had this commented out in the code as well.  Just wonder what your test results with it on and off were.
- **CompoundDecimal** - how would you feel about renaming this to RobustDouble to match with RobustGeometry and the fact that it's double based?
- **checkInvariants** - I'd like to see in the code some comments that include an example or something.  It's a little hard to follow all the bit manipulation and I think an example would help.
- **fromSum(double, double, double)** - I feel like this should be renamed to getSumError since that's what it's returning and it's not performing any modification of the input.  Same for fromDiff => getDifferenceError, fromProduct => getProductError.
- **fromDiff2x2** - I would change this so that it either returns a new CompoundDecimal or void.  Right now it's both accepting a result and returning it.
- **Cross_Product** - Needs to be camel case - probably just missed it.  More importantly, do you think we should rename this to cross to match Vector2 methods?  Also, I'm wondering if we should move this method to RobustGeometry since it's a geometric property rather than a numeric property.
- **B, C1, C2, D, and buffer** - I would move these to be method local variables so that the RobustGeometry can be thread-safe.
- **Static fields** - be sure these follow the all caps convention.
- **copyFrom(CompoundDecimal)** - any reason why there's this and a copy constructor?
- **mark()** - As far as I could tell, it looked like this was only used for method local iteration.  Can you confirm?  If so, I would remove that feature and just define local iterates.
- **Testing** - For the unit tests, are you planning to use the RobustGeometry method to determine if all the points are contained in the resulting hull?

Thanks again for all the contributions!

**Comment by mtsamis** [comment]
Thanks! Below are answers to your questions, some need further discussion.

- **strictfp** - Yeah it makes sense to use sctrictfp for this reason, and that's why I've added it there. The reason I have it commented out is to discuss:
    1) Unfortunately it reduces performance, I think somewhere up to 40-60% (when tested with cpus that otherwise have SSE and would not need in theory to have any performance hit)
    2) If we don't consider 'old CPUs' (that is those without SSE floating point instructions) then the double has the desired properties anyway
    3) That's the reason [this JEP](https://openjdk.java.net/jeps/306) is around
    4) It is probably nearly (statistically) impossible to get wrong results for this specific algorithm even with other rounding modes (although this isn't exactly a strong argument, I know).

    Now because correctness is important, one viable alternative is to update the code to use strictfp *only where needed*. The author has already done this and there are specific places where strictfp has to be used (see `imprecise double`). But this will make the code more complex in return.
- **CompundDecimal** - I think I prefer not to use the word Double in it; the fact that it uses doubles is to some extend encapsulated and not visible much outside the class. A 'Double' is a floating point representation of *some specific size*, and this class does not represent a floating point with that size. I used Decimal as it is a familiar name from BigDecimal. I have no problem to change the name, but we must find something with good meaning. RobustDecimal does not sound good to me either (don't know what it could mean). Maybe MultiprecisionDecimal?
- **checkInvariants** - Sure, I'll add in the comments an example with specific numbers, maybe some more explanations as well
- **fromSum** -  I was initially using a something of the form 'getErrorXXX()' but I did not like the word error in it. It's not important that this number is the error, It could be something else as long as it had the desired property. So what I wanted to show is that this method is not one that has to compute errors but rather one to create a representation of a CompoundDecimal *from the sum* of two values (hence the current name). We can discuss more names for this as well, I would suggest to focus on what is the meaning of this method and not how it achieves that.
- **fromDiff2x2** - Yes and the returned value is not used in this case, but I wanted to make it consistent with the sum/cross methods. The sum/cross methods sort of need this in order to be able to both return a new CompoundDecimal but also use an existing one. What do you think?
- **Cross_Product** - Yeah missed this one; Both your suggestions sound good, I'll rename to cross and move to the RobustGeometry class.
- **B, C1, C2, D, and buffer** - Good catch, I didn't think of the thread safety issue. Fortunately for 'almost all' only the B value is needed hence the performance loss isn't very big (seems to be ~10-15% for the hard cases). I'll make this change as well.
- **Static fields** - Ok, will change them
- **copyFrom(CompoundDecimal)** - I thought it could come in handy; Instead of the copy con/tor this does not allocate new space but allows to 'recycle' a CompoundDecimal. It's not used somwhere though so I can remove it as well.
- **mark()** - yeah this is used for iteration. I can remove but I would somewhat suggested to not do so. The initial code was so and I believe it is more readable in this way. You can see the original code in C which is written this way to compare. What is it that you don't like about this approach? If you wish I can revert to local index variables though, no problem.
- **Testing** - I did not quite understand :) Can you rephrase this? Do you mean if I'll add unit tests for the hull algorithms with the RobustGeometry class?

What's your opinion on not making CompoundDecimal a public class, at least for now (for the reasons I mentioned in a previous comment)?

Also to bring up one more point, even after millions of tests with computed values I did not found any that could bring the algorithm to the 'last part', the one marked
`// This case is so rare that we don't know if there are any inputs going into it`.
I recall (not sure though) I've read somewhere that even the author is not sure if such input values do exist. So none of the existing unit tests go through that part and I wanted to at least mention it. At least the code is short/straightforward and I don't think it contains bugs.

**Comment by mtsamis** [comment]
I've made some new changes as mentioned above. You can check the new commits

**Comment by wnbittle** [comment]
Great - updates look good.  To continue our discussion on the remaining items:

- **strictfp** - My preference would be to exclude it completely.  My thought here is that even without strictfp, this solution is far better than what we have today and we can always come back and add it later if we find some instances where it's failing.
- **CompoundDecimal** - Good points.  What about "Real" or just "Decimal"?  I guess i don't like the Compound part.
- **fromSum** - Are we talking about the same methods?  I was referring to the static methods that accept three doubles as arguments.  These clearly return the error of the computation - even the javadoc states that.
- **fromDiff2x2** - What I'd like to see is either: `output fn(arg, arg, arg) or void fn(arg, arg, arg, output)` instead of `output fn(arg, arg, arg, output)`.  The latter method signature implies that the forth argument could be null (as this is common with the standard Java library).
- **copyFrom(CompoundDecimal)** - Ok - no issue with this.
- **mark()** - My thinking was why would we carry around iteration state in the CompoundDecimal when we only do local iteration?  I would also disagree that it was more readable.  While reading the code, I had to examine the mark, reset, etc. methods before i even understood that it was just iterating the components.
- **Testing** - The Polygon.contains method uses the non-robust version of the side of line test and I thought that in doing so the unit tests would fail if we gave it data out of the improved hull generator code.  Was that not the case?

Regarding if the CompoundDecimal should be a public class - my opinion is to leave it package private for now.  It's easy to expose functionality later.

I wouldn't worry too much about that last case - the goal is to improve the algorithms iteratively and you are doing just that.

**Comment by mtsamis** [comment]
Alright, thanks for the extra feedback. You can also check again the new commits based on what you've said. You can also check the [new second branch](https://github.com/mtsamis/dyn4j/tree/robust_hull) which is about the actual implementations of the hull algorithms.

The new branch has the refined implementations I made based on the robust side of line test that are in general cleaner/faster (removed much code, especially for `DivideAndConquer` and optimized some few parts) and robust (passes all the randomized tests I had so far and *all the existing unit tests**)
* there are currently some tests failing, but that is because Polygon#contains is imprecise as you mentioned; I have checked with a local robust version and all tests pass indeed.

So to continue with the todo list:

- **strictfp** - Agreed, this is what I imagined as well. Done
- **CompoundDecimal** - "Real" or "Decimal" don't convey a lot of meaning for me, that's why I'm trying to find something more descriptive. Let's try some more names until we get something :)
What about "ExtendedDecimal" / "CompositeDecimal" / "LargeDecimal" or maybe even "LargeDouble" / "BigFloat" (this is accurate in a sense, as this is in reality a floating point representation with much larger but fixed value range; but I fear it can be confusing from another point of view).
- **fromSum** - Yeah we're talking about the same methods, I know it is written that it actually computes the error because the error there is *one possible value that satisfies the needed properties*. But what I'm saying is that
    1) this could be something else than the error
    2) I don't want to focus that it happens to be convenient to use the error computation for this but rather that this method generates the values for a CompoundDecimal that represents the result of an operation.

    The current names want to describe `compoundDecimalFromSum` -> `fromSum`, that is to focus on the reason they should be used and not how this is actually achieved/calculated. I wouldn't discuss this as much (after all they're just package private methods) but every time I think of renaming them to `fromSum` -> `getSumError` it feels somehow wrong to me. Maybe some alternative name? "getSumTail" (inspired by the names in the paper) / "(get)ComponentFromSum" / "(get)ErrorComponentFromSum" (lengthy).
- **fromDiff2x2** - Good advice, I made all sum/cross/fromDiff2x2 accept null as result and allocate a new sufficient decimal in that case. I also left the shorthand version that call the corresponding method with an implicit null (So both use cases are handled). The api feels better to me now, tell me if you like it as it is now.
- **mark()** - Alright, I removed the mark scheme and I can say the result is satisfying (plus gets rid of the extra field). It was initially a bit more ugly but through the mark thing I got the code to be better, so now it can finally be removed again. Thanks, done.
- **Testing** - Yeah, now I understood; I actually got to this as I mentioned in the beginning. The imprecise Polygon#contains implementation indeed causes problems in some few tests.
In my opinion it's a bit unnecessary to have a robust polygon contains method in the Polygon class as things are currently. Although the overhead won't be very important (typical Polygons should be handled by the fast case almost always) I'm not certain this is has to be changed given that probably no one will need it there.
I suggest (at least for now?) that we use a robust enough Polygon contains version only for the tests and leave Polygon class as is.

Good to see that mostly naming stuff is left to finish this! We can solve those here and I'll follow with the needed PRs. Sometime then we can have a look at the geometry.decompose package but this one sounds a bit harder to tackle. When you have a look at the hull branch feel free to discuss possible improvements.

**Comment by wnbittle** [comment]
Yeah, we're real close.  The changes to the `HullGenerator`s looks good too.

Let's firm up the last bit here (sorry been real busy lately).

- **CompoundDecimal** - ok what about Extended/Adaptive + Double, Decimal, or Precision (`ExtendedPrecision`, `AdaptivePrecision`, `ExtendedDouble`, etc. for example).

- **fromSum** - You'll probably roll your eyes (or laugh), but I like `getErrorComponentFromSum` a lot. 

- **Testing** - I think we're on the same page now.  I was suggesting that we use the robust methods in the unit tests only - the Polygon.contains method should remain as is.

Thanks again for all the contributions.

**Comment by mtsamis** [comment]
- `AdaptiveDecimal` sounds quite nice to me!
- let's go with `getErrorComponentFromSum` then, nevermind the length :grin: 
- Alright I'll implement robust Polygon.contains and friends seperately in the unit tests

Since we're done with the discussion here I'll implement those final changes and then do the PRs  (soon enough) :tada:

**Comment by wnbittle** [comment]
@mtsamis I've accepted the PRs and will do my testing today and one last review and then we can close this.
